system/metrics/sinks/log/logSink.cpp

```diff
@@ -47,57 +47,85 @@ void LogMetricSink::doCollection()
 
 void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
 {
-    std::string name = pMetric->queryName();
+    const std::string & name = pMetric->queryName();
+    __uint64 metricValue = pMetric->queryValue();
+    if (ignoreZeroMetrics && (metricValue == 0))
+        return;
+
     const auto & metaData = pMetric->queryMetaData();
+    const std::string * searchKey = &name;
+    //Create a unique id from a combination of the name and the labels so that we can check whether this metric instance has changed since the last time it was logged
+    std::string uid;
+    if (!metaData.empty())
+    {
+        uid.append(name);
+        for (auto &metaDataIt: metaData)
+            uid.append("/").append(metaDataIt.value.c_str());
+        searchKey = &uid;
+    }
+
+    auto match = alreadySeen.find(*searchKey);
+    const auto notFound = alreadySeen.end();
+    //If the values haven't changed then avoid logging an update to the logs
+    if ((match != notFound) && (match->second == metricValue))
+        return;
+
+    StringBuffer output;
+
+    StringBuffer labels;
     for (auto &metaDataIt: metaData)
     {
-        name.append(".").append(metaDataIt.value);
+        if (labels.length() > 0)
+            labels.append(",");
+        labels.appendf("{ \"name\":\"%s\", \"value\": \"%s\" }", metaDataIt.key.c_str(), metaDataIt.value.c_str());
     }
 
+    if (labels.length())
+        output.appendf("\"labels\": [%s], ", labels.str());
+
     const char *unitsStr = pManager->queryUnitsString(pMetric->queryUnits());
 
     if (pMetric->queryMetricType() != METRICS_HISTOGRAM)
     {
-        __uint64 metricValue = pMetric->queryValue();
-        if (!ignoreZeroMetrics || metricValue)
-        {
-            if (!isEmptyString(unitsStr))
-            {
-                name.append(".").append(unitsStr);
-            }
-            LOG(MCmonitorMetric, "name=%s,value=%" I64F "d", name.c_str(), metricValue);
-        }
+        output.appendf("\"value\": %" I64F "d", metricValue);
     }
     else
     {
+        output.append("\"sum\": ").append(metricValue).append(", ");
+
+        StringBuffer valueText;
         std::vector<__uint64> values = pMetric->queryHistogramValues();
-        std::vector<__uint64> limits = pMetric->queryHistogramBucketLimits();
-        size_t countBucketValues = values.size();
+        __uint64 cumulative = 0;
```

Comment:
The base histogram  does not accumulate bucket measurement counts. The Prometheus sink does since it is the required format. I left the log and file sinks to only report the actual counts per bucket to match the way the metric stored the data and possibly make it easier to extract certain counts from the log if needed. Accumulating the counts makes the output more like Prometheus. Just an overall comment.